src/advanced_slider.rs

```diff
@@ -0,0 +1,345 @@
+                    }
+                }
+
+                // Handle allowed input characters
+                druid::keyboard_types::Key::Character(string) => match string.as_str() {
+                    "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" | "." | "-" => {
+                        self.input_string.push_str(&string);
+                        self.val_text = TextLayout::from_text(format!("{}", self.input_string));
+                        self.val_text.rebuild_if_needed(ctx.text(), env);
+                        ctx.request_paint();
+                    }
+                    _ => {}
+                }
+
+                // Handle deleting chararcters of the input sting
+                druid::keyboard_types::Key::Backspace => {
+                    self.input_string.pop();
+                    self.val_text = TextLayout::from_text(format!("{}", self.input_string));
+                    self.val_text.rebuild_if_needed(ctx.text(), env);
+                    ctx.request_paint();
+                }
+
+                _ => {}
+            }
+
+            _ => {}
+        }
+    }
+
+    // Handle initialisation
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, _data: &f64, _env: &Env) {
+        match event {
+            LifeCycle::WidgetAdded => {
+                ctx.request_layout();
+                ctx.request_paint();
+            }
+            _ => {}
+        }
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &f64, data: &f64, env: &Env) {
+        if old_data != data {
+            // For the case data gets modified while in input mode
+            self.input_mode = false;
+            if self.keyboard_input_origin {
+                self.keyboard_input_origin = false;
+                ctx.request_layout();
+                ctx.request_paint();
+            } else {
+                self.val_text = TextLayout::from_text(format!("{:.*}", self.signif_dig, data));
+                self.val_text.rebuild_if_needed(ctx.text(), env);
+                ctx.request_layout();
+                ctx.request_paint();
+            }
+            
+        }
+    }
+
+    fn layout(
+        &mut self,
+        _layout_ctx: &mut LayoutCtx,
+        _bc: &BoxConstraints,
+        _data: &f64,
+        _env: &Env,
+    ) -> Size {
+        Size::new(124.0, 24.0)
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, data: &f64, _env: &Env) {
+        // Handle in which mode to draw the widget
+        if self.input_mode {
+            let rounded_box = RoundedRect::new(2.0, 2.0, 122.0, 22.0, 2.0);
+            ctx.fill(rounded_box, &Color::rgb8(50, 50, 50));
+            ctx.stroke(rounded_box, &Color::rgb8(80, 80, 80), 1.0);
+
+            // Center Text and draw it
+            let text_width = self.val_text.layout_metrics().size.width;
+            self.val_text.draw(
+                ctx,
+                Point::new(62.0 - (text_width / 2.0), 2.0 + self.text_offset),
+            );
```

Comment:
Looks great.

Curious whether having this outside if-else would make a difference?

Reply:
I think it would not, so I'll change that as well.
Thanks for the hint.